File: push_swap_prototype/lab/push_swap/src/op_push.c
#include "models.h"

void	save_op(t_ps *data, t_op op);

int	next_up(t_stack *stk, int index);
bool	is_full(t_stack *stk);
int	next_down(t_stack *stk, int index);

void	push(t_stack *src, t_stack *dest)
{
	int	dest_i;

	if (is_full(dest))
		return ;
	dest_i = next_up(dest, dest->top);
	dest->stack[dest_i] = src->stack[src->top];
	dest->top = dest_i;
	src->stack[src->top] = 0;
	src->top = next_down(src, src->top);
}

void	push_a(t_ps *data)
{
	push(&data->b, &data->a);
	if (data->writing_mode)
		save_op(data, pa);
}

void	push_b(t_ps *data)
{
	push(&data->a, &data->b);
	if (data->writing_mode)
		save_op(data, pb);
}

File: push_swap_prototype/src/internal/stack/op_push.go
package stack


func push(src *Stack, dest *Stack) {
	if IsFull(dest) {
		return
	}
	dest_i := NextUp(dest, dest.Top)
	dest.Stack[dest_i] = src.Stack[src.Top]
	dest.Top = dest_i
	src.Stack[src.Top] = 0
	src.Top = NextDown(src, src.Top)
}

func PushA(ps *SortingState) {
	// Enhanced logging and validation
	LogOperation("PushA", ps)
	
	// Validate stack state before operation
	result := ValidateStackState(ps, "before PushA")
	if !result.IsValid {
		LogError("Stack validation failed before PushA:")
		for _, err := range result.Errors {
			LogError("  - %s", err)
		}
		// Continue with operation but log the issue
	}
	
	// Perform the push operation
	push(ps.B, ps.A)
	
	// Validate stack state after operation
	result = ValidateStackState(ps, "after PushA")
	if !result.IsValid {
		LogError("Stack validation failed after PushA:")
		for _, err := range result.Errors {
			LogError("  - %s", err)
		}
	}
	
	// Log warnings if any
	for _, warning := range result.Warnings {
		LogWarning("  - %s", warning)
	}
	
	// Save operation if in writing mode
	if ps.WritingMode {
		SaveOp(ps, PA)
	}
	
	LogOperationResult("PushA", ps)
}

func PushB(ps *SortingState) {
	// Enhanced logging and validation
	LogOperation("PushB", ps)
	
	// Validate stack state before operation
	result := ValidateStackState(ps, "before PushB")
	if !result.IsValid {
		LogError("Stack validation failed before PushB:")
		for _, err := range result.Errors {
			LogError("  - %s", err)
		}
		// Continue with operation but log the issue
	}
	
	// Perform the push operation
	push(ps.A, ps.B)
	
	// Validate stack state after operation
	result = ValidateStackState(ps, "after PushB")
	if !result.IsValid {
		LogError("Stack validation failed after PushB:")
		for _, err := range result.Errors {
			LogError("  - %s", err)
		}
	}
	
	// Log warnings if any
	for _, warning := range result.Warnings {
		LogWarning("  - %s", warning)
	}
	
	// Save operation if in writing mode
	if ps.WritingMode {
		SaveOp(ps, PB)
	}
	
	LogOperationResult("PushB", ps)
}